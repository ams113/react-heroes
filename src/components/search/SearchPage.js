import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchPage = ( { history } ) => {

    
    const location = useLocation();
    const { q = '' } = queryString.parse( location.search );

    const [ formValues, handleInputChange ] = useForm({ searchText: q });
    const { searchText } = formValues;

  
    
    // const heroesFitered = getHeroesByName( searchText );
    const heroesFitered = useMemo(() => getHeroesByName( q ), [q])
    const handleSearch = ( e ) => {
        e.preventDefault()

        history.push(`?q=${ searchText }`);
    };
    
    return (
        <div>
            <div className="d-none d-sm-block">
                <h1>Search heroes</h1>
                <hr />
            </div>

            <div className="row animate__animated animate__fadeIn">

                <div className="col-md-4">

                    <form onSubmit={ handleSearch }>

                    <div className="d-grid gap-2 mb-3">
                        <input 
                            type="text" 
                            name="searchText"
                            placeholder="Find you hero"
                            className="form-control"
                            autoComplete="off"
                            value={ searchText }
                            onChange={ handleInputChange }
                        />

                        <button 
                            type="submit"
                            className="btn btn-outline-primary"
                        >
                            Search
                        </button>
  
                    </div>
                        
                    </form>

                </div>

                <div className="col-md-6 offset-lg-1">
                    
                    <h4>Results:</h4>
                    
                    {
                        ( q === '' ) &&
                        <div className="alert alert-secondary">
                            Search a hero
                        </div>
                    }
                    {
                        ( q !== ''  && heroesFitered.length === 0 ) &&
                        <div className="alert alert-danger">
                            There is not a hero with <strong>{ q }</strong>
                        </div>
                    }
                    {
                        heroesFitered.map( hero => (
                            <HeroCard   
                                key={ hero.id } 
                                { ...hero } 
                            />
                        ))
                    }

                </div>

            </div>
        </div>
    )
}
