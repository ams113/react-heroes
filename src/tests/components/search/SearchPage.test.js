import React from 'react'
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchPage } from '../../../components/search/SearchPage';


describe('Testing Components: SearchPage', () => {


    test('should render correctly with default values', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={ SearchPage } />
            </MemoryRouter>
        );
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-secondary').text().trim() ).toBe('Search a hero');

    });

    test('should render Batman Card and input with queryString value', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={ SearchPage } />
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value') ).toBe('batman');
        expect( wrapper.find('HeroCard').exists() ).toBe(true);
        expect( wrapper ).toMatchSnapshot();

    });

    test('should render alert error if the hero is not found', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=foo-man']}>
                <Route path="/search" component={ SearchPage } />
            </MemoryRouter>
        );

        expect( wrapper.find('.alert-danger').text().trim() ).toBe('There is not a hero with foo-man');
        expect( wrapper ).toMatchSnapshot();

    });

    test('should call history push', () => {

        const param = 'batman';
        const history = {
            push: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=' + param]}>
                <Route 
                    path="/search"
                    component={ 
                        () => <SearchPage history={ history }/> 
                    } />
            </MemoryRouter>
        );

        wrapper.find('button').simulate('change', {
            target: {
                name: 'searchText',
                value: param
            }
        });

        /* wrapper.find('form').prop('onSubmit')(
            { preventDefault: jest.fn() }
        ); */
        wrapper.find('form').simulate('submit')

        expect( history.push ).toBeCalledWith('?q=' + param );

    })
    
    
})
