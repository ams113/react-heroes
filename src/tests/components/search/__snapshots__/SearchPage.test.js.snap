// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Components: SearchPage should render Batman Card and input with queryString value 1`] = `
<div>
  <div
    className="d-none d-sm-block"
  >
    <h1>
      Search heroes
    </h1>
    <hr />
  </div>
  <div
    className="row animate__animated animate__fadeIn"
  >
    <div
      className="col-md-4"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="d-grid gap-2 mb-3"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="searchText"
            onChange={[Function]}
            placeholder="Find you hero"
            type="text"
            value="batman"
          />
          <button
            className="btn btn-outline-primary"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-md-6 offset-lg-1"
    >
      <h4>
        Results:
      </h4>
      <div
        className="col"
      >
        <div
          className="card md-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row g-0"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="img-fluid"
                src="./assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  More info ...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing Components: SearchPage should render alert error if the hero is not found 1`] = `
<div>
  <div
    className="d-none d-sm-block"
  >
    <h1>
      Search heroes
    </h1>
    <hr />
  </div>
  <div
    className="row animate__animated animate__fadeIn"
  >
    <div
      className="col-md-4"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="d-grid gap-2 mb-3"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="searchText"
            onChange={[Function]}
            placeholder="Find you hero"
            type="text"
            value="foo-man"
          />
          <button
            className="btn btn-outline-primary"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-md-6 offset-lg-1"
    >
      <h4>
        Results:
      </h4>
      <div
        className="alert alert-danger"
      >
        There is not a hero with 
        <strong>
          foo-man
        </strong>
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing Components: SearchPage should render correctly with default values 1`] = `
<div>
  <div
    className="d-none d-sm-block"
  >
    <h1>
      Search heroes
    </h1>
    <hr />
  </div>
  <div
    className="row animate__animated animate__fadeIn"
  >
    <div
      className="col-md-4"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="d-grid gap-2 mb-3"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="searchText"
            onChange={[Function]}
            placeholder="Find you hero"
            type="text"
            value=""
          />
          <button
            className="btn btn-outline-primary"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-md-6 offset-lg-1"
    >
      <h4>
        Results:
      </h4>
      <div
        className="alert alert-secondary"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
